src/combinatorics/catalan.lean

```diff
@@ -129,3 +141,54 @@ by norm_num [catalan_eq_central_binom_div, nat.central_binom, nat.choose]
 
 lemma catalan_three : catalan 3 = 5 :=
 by norm_num [catalan_eq_central_binom_div, nat.central_binom, nat.choose]
+
+namespace tree
+open_locale tree
+
+/-- Given two finsets, find all trees that can be formed with
+  left child in `a` and right child in `b` -/
+@[reducible] def pairwise_node (a b : finset (tree unit)) : finset (tree unit) :=
+(a ×ˢ b).map ⟨λ x, x.1 △ x.2, λ ⟨x₁, x₂⟩ ⟨y₁, y₂⟩, λ h, by simpa using h⟩
+
+/-- A finset of all trees with `n` nodes. See `mem_trees_of_nodes_eq` -/
+def trees_of_num_nodes_eq : ℕ → finset (tree unit)
+| 0 := {nil}
+| (n+1) := (finset.nat.antidiagonal n).attach.bUnion $ λ ijh,
```

Comment:
@YaelDillies, can you see a way to use `disj_Union` here instead?

Reply:
I thought about it but it doesn't seem obvious at all.